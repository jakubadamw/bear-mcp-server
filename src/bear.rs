use crate::Result;

use serde::{Deserialize, Serialize};
use sqlx::SqlitePool;
use std::path::PathBuf;

#[derive(sqlx::Type)]
#[sqlx(transparent)]
pub struct AppleCoreDateTime(f64);

impl From<AppleCoreDateTime> for chrono::DateTime<chrono::Utc> {
    fn from(value: AppleCoreDateTime) -> Self {
        chrono::NaiveDateTime::new(
            chrono::NaiveDate::from_ymd_opt(2001, 1, 1).expect("must succeed"),
            chrono::NaiveTime::from_hms_opt(0, 0, 0).expect("must succeed"),
        )
        .and_utc()
            + chrono::Duration::microseconds((value.0 * 1_000_000.0f64) as i64)
    }
}

#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct Note {
    #[sqlx(try_from = "AppleCoreDateTime")]
    creation_date: chrono::DateTime<chrono::Utc>,
    subtitle: Option<String>,
    text: Option<String>,
    title: Option<String>,
    id: String,
}

#[derive(sqlx::FromRow)]
struct Tag {
    title: String,
}

pub struct BearDatabase {
    pool: SqlitePool,
}

impl BearDatabase {
    pub async fn new(db_path: PathBuf) -> Result<Self> {
        let database_url = format!("sqlite://{}?mode=ro", db_path.display());
        let pool = SqlitePool::connect(&database_url).await?;
        Ok(Self { pool })
    }

    pub async fn list_notes(&self, query: Option<&str>, tag: Option<&str>) -> Result<Vec<Note>> {
        let search_pattern: Option<String> = query.map(|query| format!("%{query}%"));
        let search_pattern_str = search_pattern.as_deref();

        Ok(sqlx::query_as!(
            Note,
            "SELECT
                creation_date AS 'creation_date!: AppleCoreDateTime',
                subtitle,
                text,
                title,
                id AS 'id!'
            FROM (
                SELECT
                    notes.ZCREATIONDATE AS creation_date,
                    notes.ZSUBTITLE AS subtitle,
                    notes.ZTEXT AS text,
                    notes.ZTITLE AS title,
                    notes.ZUNIQUEIDENTIFIER AS id,
                    (
                        CASE
                            WHEN ? IS NULL THEN 1.0
                            WHEN notes.ZTITLE = ? THEN 0.75
                            WHEN notes.ZTITLE LIKE ? THEN 0.5
                            WHEN notes.ZTEXT LIKE ? THEN 0.25
                            ELSE 0.0
                        END
                    ) AS text_search_score
                FROM 
                    ZSFNOTE notes
                LEFT JOIN 
                    Z_5TAGS note_tags ON notes.Z_PK = note_tags.Z_5NOTES
                INNER JOIN 
                    ZSFNOTETAG tags ON tags.Z_PK = note_tags.Z_13TAGS
                WHERE 
                    (? IS NULL OR tags.ZTITLE = ?)
                    AND text_search_score > 0
            )
            LIMIT
                10",
            search_pattern_str,
            query,
            search_pattern_str,
            search_pattern_str,
            tag,
            tag
        )
        .fetch_all(&self.pool)
        .await?)
    }

    pub async fn list_tags(&self) -> Result<Vec<String>> {
        Ok(sqlx::query_as!(
            Tag,
            "SELECT
                ZTITLE AS 'title!'
            FROM
                ZSFNOTETAG"
        )
        .fetch_all(&self.pool)
        .await?
        .into_iter()
        .map(|tag| tag.title)
        .collect())
    }
}
