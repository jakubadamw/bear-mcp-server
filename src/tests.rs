use std::path::PathBuf;

use crate::bear::BearDatabase;

#[rstest::rstest]
#[case(0, None, Some("first_tag"))]
#[case(1, None, Some("second_tag"))]
#[case(2, Some("Lorem"), None)]
#[case(3, Some("Test note with multiple tags"), None)]
#[case(4, Some("Test note with another tag"), Some("second_tag"))]
#[tokio::test]
async fn test_list_notes(
    #[case] index: usize,
    #[case] query: Option<&str>,
    #[case] tag: Option<&str>,
) {
    let path = PathBuf::from(env!("CARGO_MANIFEST_PATH"))
        .parent()
        .expect("must have a parent")
        .join("test_data")
        .join("database.sqlite");

    let bear_database = BearDatabase::new(path).await.expect("must succeed");
    let result = bear_database
        .list_notes(query, tag)
        .await
        .expect("must succeed");
    let snapshot_file_name = format!("test_list_notes_{index:02}");
    insta::assert_json_snapshot!(snapshot_file_name, result);
}
