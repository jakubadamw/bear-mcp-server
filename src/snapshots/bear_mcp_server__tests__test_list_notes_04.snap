---
source: src/tests.rs
expression: result
---
[]
